### ISSUE TITLE
Very Slow MxNet distributed training 

### ISSUE THREAD

**Issue by salemmohammed** [dmlc/mxnet#5422]
I am working on the example training_mnist.py and I notice that it's very slow when I run the distributed training. Also, cross entropy value gave me Nan. When I ran it on my single machine, everything works fine but I believe it has slow speed training per sample.

Here is the result for training on single machine:

`python train_mnist.py --network mlp`

INFO:root:Epoch[0] Batch [100]	Speed: 40726.16 samples/sec	Train-cross-entropy=1.054846
INFO:root:Epoch[0] Batch [200]	Speed: 41420.49 samples/sec	Train-cross-entropy=0.544122
INFO:root:Epoch[0] Batch [300]	Speed: 41170.83 samples/sec	Train-cross-entropy=0.471626
INFO:root:Epoch[0] Batch [400]	Speed: 39867.68 samples/sec	Train-cross-entropy=0.437326
INFO:root:Epoch[0] Batch [500]	Speed: 40247.01 samples/sec	Train-cross-entropy=0.414339
INFO:root:Epoch[0] Batch [600]	Speed: 44374.19 samples/sec	Train-cross-entropy=0.396647
INFO:root:Epoch[0] Batch [700]	Speed: 40458.16 samples/sec	Train-cross-entropy=0.377359
INFO:root:Epoch[0] Batch [800]	Speed: 39988.00 samples/sec	Train-cross-entropy=0.360503
INFO:root:Epoch[0] Batch [900]	Speed: 40203.55 samples/sec	Train-cross-entropy=0.358936
INFO:root:Epoch[0] Train-cross-entropy=0.350466
INFO:root:Epoch[0] Time cost=1.476

I changed map.py file to:
`
def get_symbol(num_classes=10, **kwargs):
    data = mx.symbol.Variable('data')
    data = mx.sym.Flatten(data=data)
    fc3  = mx.symbol.FullyConnected(data = data, name='fc3', num_hidden=num_classes)
    mlp  = mx.symbol.SoftmaxOutput(data = fc3, name = 'softmax')
    return mlp
`

Here is the result for distributed training.
`../../tools/launch.py -n 3 -H hosts --sync-dst-dir /tmp/mxnet python train_mnist.py --network mlp --kv-store dist_sync`

INFO:root:Epoch[0] Batch [100]	Speed: 149.58 samples/sec	Train-cross-entropy=nan
INFO:root:Epoch[0] Batch [100]	Speed: 149.42 samples/sec	Train-cross-entropy=nan
INFO:root:Epoch[0] Batch [100]	Speed: 149.78 samples/sec	Train-cross-entropy=nan

Finally, Does TensorFlow faster than Mxnet ?

Sincerely

**Comment by zzmig** [comment]
see https://github.com/dmlc/mxnet/issues/5158 . This may help you.

**Comment by salemmohammed** [comment]
@philipskokoh 
I got why the distributed training is slow but I still wondering why the cross entropy is Nan.
May you please give me a hint ?

**Comment by salemmohammed** [comment]
@piiswrong 

I ran the Cifar dataset and the speed is also slow. I do not know why! Does anybody has this experience ?


`INFO:root:Epoch[0] Batch [20]	Speed: 2471.16 samples/sec	Train-accuracy=0.100190
INFO:root:Epoch[0] Batch [20]	Speed: 2576.56 samples/sec	Train-accuracy=0.098667
INFO:root:Epoch[0] Batch [20]	Speed: 2249.62 samples/sec	Train-accuracy=0.099905
INFO:root:Epoch[0] Train-accuracy=0.102923
INFO:root:Epoch[0] Time cost=6.688
INFO:root:Epoch[0] Train-accuracy=0.104154
INFO:root:Epoch[0] Time cost=6.691
INFO:root:Epoch[0] Train-accuracy=0.095231
INFO:root:Epoch[0] Time cost=7.166
INFO:root:Epoch[1] Batch [20]	Speed: 2786.97 samples/sec	Train-accuracy=0.099619
INFO:root:Epoch[1] Batch [20]	Speed: 2803.88 samples/sec	Train-accuracy=0.098571
INFO:root:Epoch[1] Batch [20]	Speed: 2781.56 samples/sec	Train-accuracy=0.100000`
